hack/generator/pkg/astmodel/resource.go

```diff
@@ -19,71 +19,109 @@ type ResourceType struct {
 func NewResourceType(specType Type, statusType Type) *ResourceType {
-	return &ResourceType{specType, statusType, false}
+	return &ResourceType{
+		specType, statusType, false, make(map[Flag]struct{})}
 }
 
 // assert that ResourceType implements Type
 var _ Type = &ResourceType{}
 
 // SpecType returns the type used for specificiation
-func (definition *ResourceType) SpecType() Type {
-	return definition.spec
+func (resource *ResourceType) SpecType() Type {
+	return resource.spec
 }
 
 // StatusType returns the type used for current status
-func (definition *ResourceType) StatusType() Type {
-	return definition.status
+func (resource *ResourceType) StatusType() Type {
+	return resource.status
 }
 
 // AsType converts the ResourceType to go AST Expr
-func (definition *ResourceType) AsType(_ *CodeGenerationContext) ast.Expr {
+func (resource *ResourceType) AsType(_ *CodeGenerationContext) ast.Expr {
 	panic("a resource cannot be used directly as a type")
 }
 
 // Equals returns true if the other type is also a ResourceType and has Equal fields
-func (definition *ResourceType) Equals(other Type) bool {
-	if definition == other {
+func (resource *ResourceType) Equals(other Type) bool {
+	if resource == other {
 		return true
 	}
 
+	if resource == nil {
+		return false
+	}
+
 	if otherResource, ok := other.(*ResourceType); ok {
-		return TypeEquals(definition.spec, otherResource.spec) &&
-			TypeEquals(definition.status, otherResource.status) &&
-			definition.isStorageVersion == otherResource.isStorageVersion
+
+		if otherResource == nil {
+			return false
+		}
+
+		if !TypeEquals(resource.spec, otherResource.spec) ||
+			!TypeEquals(resource.status, otherResource.status) ||
+			!resource.isStorageVersion == otherResource.isStorageVersion ||
+			len(resource.flags) != len(otherResource.flags) {
+			return false
+		}
+
+		for f := range resource.flags {
+			if _, ok := otherResource.flags[f]; !ok {
+				return false
+			}
+		}
+
+		return true
 	}
 
 	return false
 }
 
 // References returns the types referenced by Status or Spec parts of the resource
-func (definition *ResourceType) References() TypeNameSet {
-	spec := definition.spec.References()
+func (resource *ResourceType) References() TypeNameSet {
+	spec := resource.spec.References()
 
 	var status TypeNameSet
-	if definition.status != nil {
-		status = definition.status.References()
+	if resource.status != nil {
+		status = resource.status.References()
 	}
 
 	return SetUnion(spec, status)
 }
 
 // MarkAsStorageVersion marks the resource as the Kubebuilder storage version
-func (definition *ResourceType) MarkAsStorageVersion() *ResourceType {
-	result := *definition
+func (resource *ResourceType) MarkAsStorageVersion() *ResourceType {
+	result := resource.copy()
 	result.isStorageVersion = true
-	return &result
+	return result
+}
+
+// AddFlag includes the specified flag on the resource
+func (resource *ResourceType) AddFlag(flag Flag) *ResourceType {
```

Comment:
`WithFlag` maybe? (I think we use `WithXxx` elsewhere) `AddFlag` makes it sound more like a mutating operation to me.

Reply:
`WithFlag()` implies that any other flags are discarded, which we don't want - flags accumulate - which is why I went with `AddFlag()`. 

Agree that it's not ideal, do you have a better suggestion?

Reply:
We use `With` in a number of other places in the code where it's not discarded.

`PropertyDefinition`: `WithValidation`, `WithTag`.
`ObjectType`: `WithProperty` `WithProperties`, `WithFunction`, `WithInterface`
(There's probably more but I didn't bother looking further)

I don't think `With` implies _replacement of the whole property_ - it only implies full replacement if the thing you're operating on is a single value. At least to me, `With` implies that it will return an object with the field requested. If that's in the context of a single value (a string, a type, etc) then yes, it's replacement, but if it's in the context of a map or slice, it's just addition (other elements of the slice/map are preserved, not destroyed).

Reply:
You've convinced me. I'll rename it to `WithFlag()`.